components/tls/tls-client-async/src/lib.rs

```diff
@@ -74,165 +72,163 @@ impl Future for ConnectionFuture {
     let (mut rx_sender, rx_receiver) = mpsc::channel(1 << 14);
-    let (close_send, mut close_recv) = oneshot::channel();
 
-    let conn = TlsConnection::new(tx_sender, rx_receiver, close_send);
+    let conn = TlsConnection::new(tx_sender, rx_receiver);
 
     let fut = async move {
         client.start().await?;
+        let mut notify = client.get_notify().await?;
 
         let (mut server_rx, mut server_tx) = socket.split();
 
         let mut rx_tls_buf = [0u8; RX_TLS_BUF_SIZE];
         let mut rx_buf = [0u8; RX_BUF_SIZE];
 
+        let mut handshake_done = false;
         let mut client_closed = false;
         let mut server_closed = false;
 
         let mut sent = Vec::with_capacity(1024);
         let mut recv = Vec::with_capacity(1024);
 
         let mut rx_tls_fut = server_rx.read(&mut rx_tls_buf).fuse();
+        // We don't start writing application data until the handshake is complete.
+        let mut tx_recv_fut: Fuse<Next<'_, mpsc::Receiver<Bytes>>> = Fuse::terminated();
+
+        // Runs both the tx and rx halves of the connection to completion.
+        // This loop does not terminate until the *SERVER* closes the connection and
+        // we've processed all received data. If an error occurs, the `TlsConnection`
+        // channels will be closed and the error will be returned from this future.
+        'conn: loop {
+            // Write all pending TLS data to the server.
+            if client.wants_write() && !client_closed {
+                #[cfg(feature = "tracing")]
+                trace!("client wants to write");
+                while client.wants_write() {
+                    let _sent = client.write_tls_async(&mut server_tx).await?;
+                    #[cfg(feature = "tracing")]
+                    trace!("sent {} tls bytes to server", _sent);
+                }
+                server_tx.flush().await?;
+            }
 
-        'outer: loop {
-            select_biased! {
-                read_res = &mut rx_tls_fut => {
-                    let received = read_res?;
+            // Forward received plaintext to `TlsConnection`.
+            while !client.plaintext_is_empty() {
+                let read = client.read_plaintext(&mut rx_buf)?;
+                recv.extend(&rx_buf[..read]);
+                // Ignore if the receiver has hung up.
+                _ = rx_sender
+                    .send(Ok(Bytes::copy_from_slice(&rx_buf[..read])))
+                    .await;
+                #[cfg(feature = "tracing")]
+                trace!("forwarded {} plaintext bytes to conn", read);
+            }
+
+            if !client.is_handshaking() && !handshake_done {
+                #[cfg(feature = "tracing")]
+                debug!("handshake complete");
+                handshake_done = true;
+                // Start reading application data that needs to be transmitted from the `TlsConnection`.
+                tx_recv_fut = tx_receiver.next().fuse();
```

Comment:
Seems like with the current logic it is possible to silently lose some tx `data`.

1. Consider if some tx `data` is available at this point, then `tx_recv_fut` should resolve to that `data`.

2. Consider if in the `select_biased` loop below only the `rx_tls_fut` branch will be executed because there is some incoming data from the server (`select_biased!` executes only one branch iiuc). This means that the `tx_recv_fut` branch will **not** be executed and the `data` will **not** be sent.

3. In the next loop iteration `tx_recv_fut` on this line will be set to a future which resolves to something else, e.g. `None`. This will result in `data` not getting sent.